### DISCUSSION TITLE
v2: Transform XML with environment variables

### DISCUSSION THREAD

**Discussion by eirikb** [parcel-bundler/parcel#5844] [Q&A] [answered]
I'm trying to transform Office 365 Add-in manifest file (**manifest.xml**) with environment variables ,e.g., `HOME_URL` will change based on local install / production install.  
I understand this isn't something Parcel by itself should handle, but I'm not sure how to proceed.

What I've tried so far:
  * `.parcelrc`: 
```json
{
  "extends": ["@parcel/config-default"],
  "transformers": {
    "*.xml": ["@parcel/transformer-posthtml"]
  }
}
```
  * `posthtml.config.js`:
```js
require("dotenv").config();  // <- had to do this hack, parcel doesn't seem to pass the env (yes I've tried specified properties (security thing)).
module.exports = {
  plugins: {
    "posthtml-expressions": {
      locals: process.env,
    },
  },
};
```
  *  `./node_modules/.bin/parcel manifest.xml`

The approach above almost works. I get a manifest file with the environment variables.  
But `posthtml` seems to alter the XML too much (e.g., casing on attributes). I've tried searching for solutions to this, but haven't found anything good yet, and I suspect I might run into other issues when using posthtml.
Can I use another transformer? Mustache-like transformer could work, as I basically want string + environment variable templating.
Are there any other solutions?

Thanks

**Comment by devongovett** [comment, marked as answer]
I would write a custom transformer for XML if I were you. HTML syntax is too dissimilar from XML these days. There are plenty of XML parsers on npm you could use. See [the docs](https://v2.parceljs.org/plugin-system/transformer/) for details on how to write a transformer, and the source code for some of our existing transformers. Note that only the `transform` method is required.

**Reply by eirikb** [reply]
I believe the Office Add-in stuff could be even more picky about the XML than usual.
Hence why I considered a mustache transformer.
Would that be a good solution? I can find one such plugin for Parcel 1, but none for v2.

**Reply by eirikb** [reply]
Hi @devongovett 

I've tried to make a rudimentary plugin, and it seems to work fine.
It looks like this:
```js
const { Transformer } = require('@parcel/plugin');
const mustache = require('mustache');
const dotenv = require('dotenv');

module.exports = new Transformer({
  async transform({ asset }) {
    const env = dotenv.config().parsed;
    const code = await asset.getCode();
    const rendered = mustache.render(code, env);
    asset.setCode(rendered);
    return [asset];
  }
})
```

I have some questions:
  1. The only way to consume this is to both add it to `devDependencies` and add a **.parcelrc**-file like this: ?
```json
{ 
  "extends": "@parcel/config-default",
  "transformers": { "*.xml": [ "@eirikb/parcel-transformer-mu" ] }
}
```
I vaguely remember in Parcel 1 it would suffice to add a plugin as a dependency, then the plugin would be called based on extension or something.
  
  2. Is it possible for my plugin to rename the file? At the moment it's called **.xml** so it's not a big problem, but it could be nice to support renaming **.mustache.xml** or **.xml.mustache** to **.xml**.
  3. Not a technical question, but would it be odd for the plugin to look for similarly named **.json**-files in companion to the template file? E.g., **manifest.json** when using **manifest.xml** (or **manifest.xml.mustache**) as template?

Lastly a completely different question (could be another discussion):
I want to move a **icon.png** file to **dist** without hashed name, how would I go about that?
  * By default  **.png** doesn't have a transformer, so it can't be used as entry.
    I want to use entry because it will be used by Outlook, referenced in the manifest file, it won't be used in the actual source code.
  * Using **.parcelrc** and `@parcel/transformer-raw` for **.png** has `asset.isIsolated = true;`, which seems to force a hashed name (although entries usually don't have it, and I don't want it).
  * I saw two other transformers, `inline` and `image`, although they don't come with current parcel version, and I'm not sure if they are on npm.

**Reply by mischnic** [reply]
> The only way to consume this is to both add it to devDependencies and add a .parcelrc-file like this: ?

Yes

> Is it possible for my plugin to rename the file

You'd need to add a namer plugin to rename the bundle (because the asset name itself doesn't directly determine the bundle name)

(Unless you call the input file `manifest.mustache` and set `asset.type = 'xml';` in the transformer)

> but would it be odd for the plugin to look for similarly named .json-files in companion to the template file? E.g., manifest.json when using manifest.xml (or manifest.xml.mustache) as template?

(For example the webextension transformer looks into a folder called `_locale` (but the browsers also do this, so this is a standard anyway)).

> I want to move a icon.png file to dist without hashed name, how would I go about that?

Passing `isEntry: true` should achieve that.

And with `addURLDependency`, it should not throw an error that png doesn't have a transformer.

> and I'm not sure if they are on npm.

They are. And they are also still part of the default config: https://github.com/parcel-bundler/parcel/blob/v2/packages/configs/default/index.json

**Reply by eirikb** [reply]
Thanks for the reply @mischnic  :smile: 

I will publish my plugin when I get the time. I would appreciate Mustache/Handlebars being part of Parcel v2 though, or something similar.
I tried the `asset.type = 'xml'` solution, but it only works if the file is called **.mustache**, I imagine people would call the file **.xml.mustache**, or even **.mustache.xml** to get automatic file association while still hinting it's a template. 
I could be wrong though. In my case I'll just call it **manifest.xml** and be done with it.

About **icon.png**, where do I pass `isEntry: true`? I don't have a plugin for this, just normal parcel.
I tried `import 'url:./icon.png'` but it will still be hashed - this is expected, I know the `url`-part is used to get the actual URL of the file. But I need the URL in the manifest. It should probably be injected there in my Mustache plugin (I'm not sure how), but for now just not hashing would solve the problem.

About packages not in npm: I had some issues with plugins being installed when running nightly, not sure why. E.g., if I install nightly of **parcel** now I get **603**, but all the other packages are **605** and it will start installing those. I could perhaps solve this by setting specific versions in my lock file. For now I solved it by changing back to **beta**. 
Reason I changed to nightly in the first place was because I got the `regeneratorRuntime is not defined` when compiling TS (without `browserslist`, but with `"target":"es5"` in tsconfig, and I don't get this error with nightly).

**Reply by mischnic** [reply]
> I don't have a plugin for this

I though this was somehow contained in the manifest. There isn't really a way to achieve that.

**Reply by eirikb** [reply]
It's going to be referenced from the manifest. I just imagined making it dynamically insert correct file name was beyond me.
That's also something I'm using the environment variables for. Now it looks something like this in **manifest.xml**:
```xml
<bt:Image id="icon" DefaultValue="{{{ HOME_URL }}}/icon.png"/>
```

`HOME_URL` is set in env, since it must be an absolute path to where the Outlook Add-in is hosted eventually (some CDN). Yes it will "point to itself", weird.

**Reply by mischnic** [reply]
In the transformer, you could do something like `asset.addURLDependency("icon.png", {isEntry: true})`

**Reply by eirikb** [reply]
I see, thanks.
But then it wouldn't be very 'generic'. Perhaps I could look for file names and resolve them dynamically.
In that case I probably should go the XML parser path, instead of Mustache.
E.g., use `xml2js` to convert to object, traverse it, look for text/attributes such as `url:./icon.png` and then resolve files, and `env:HOST_URL`.
I'll think about it. Thanks :+1:

**Reply by eirikb** [reply]
Sorry to bring this up again, but a small update:
I noticed I could do this in mustache: `{{#url}}icon.png{{/url}}`, and in my mustache `view`:
```js
view.url = () => text => asset.addURLDependency(text, {
    isAsync: false,
    isUrl: false,
  });
```

This ensures the **icon.png**-file is bundled, but all I get back is a hash.  
Later I found this: https://github.com/parcel-bundler/parcel/discussions/5393 .
After successfully writing my own packager I noticed I could just use **@parcel/packager-raw-url**.

My final `.parcelrc`now looks like this:
```json
{
  "extends": "@parcel/config-default",
  "transformers": {
    "*.xml": [
      "@eirikb/parcel-transformer-mustache"
    ]
  },
  "packagers": {
    "*.xml": "@parcel/packager-raw-url"
  }
}
```
I suspect I would need to tell users of my npm package to use a similar setup in their `.parcelrc`-file.  
And if they want to rename **.mustache**-files they need a custom namer (a generic regex-like re-name namer would be nice).  
Would also be cool if I could implement all three phases in a single npm package for users to add, and then they could have it run for any file called **.mustache**.

**Reply by mischnic** [reply]
You could create a config package like `@parcel/config-webextension`: https://github.com/parcel-bundler/parcel/blob/v2/packages/configs/webextension/index.json

Then the users only need to add one line: https://v2.parceljs.org/recipes/web-extension/

**Reply by eirikb** [reply]
But they could only extend one of these right? So if there are other plugins doing the same as me that wouldn't work?

**Reply by mischnic** [reply]
No, you can specify multiple configs: https://github.com/parcel-bundler/parcel/blob/v2/packages/core/integration-tests/test/integration/config-merging/.parcelrc

**Reply by eirikb** [reply]
Oh I did not know this. That's great!

**Reply by eirikb** [reply]
Hi again.
I've been experimenting with writing a Namer. I already solved my initial problem, but I want to make my plugin a bit more attractive, and the config solution above would do that.  
But I'm struggling a bit with the Namer. This is what I got so far:
```js
export default new Namer({
  name({ bundle, bundleGraph }) {
    const bundleGroup = bundleGraph.getBundleGroupsContainingBundle(bundle)[0];
    const entryFilePath = nullthrows(
      bundle
        .getEntryAssets()
        .find(a => a.id === bundleGroup.entryAssetId)
    ).filePath;
    return path.basename(entryFilePath).replace(/\.mustache/i, '');
  },
});
```

All it is suppose to do is rename both **manifest.mustache.xml** and **manifest.xml.mustache** to **manifest.xml**.  
But I'm getting this:
> @eirikb/parcel-namer-mustache: Destination name manifest2.xml extension does not match bundle type "mustache"

So I tried forcing `asset.type` in my Transformer to `'xml'`, but then it complains about not finding a transformer for **manifest.xml.xml**, I'm also not sure I should do anything in my Transformer, the re-naming should be part of Namer.  
Also tried setting `bundle.type = 'xml'`, but it's a read-only property.

**Reply by eirikb** [reply]
Hi @mischnic and thanks for all the help.
I've decided to publish a plugin based on our discussions, you can find it here: https://www.npmjs.com/package/@eirikb/parcel-transformer-mustache
It is split into three packages: transformer, namer and config.  
The config is just a convenience with default setup, while transformer can be used stand-alone.
I never managed to do the namer with .mustache, only .mustache.*, but that's fine.

Hope I've set it up like you guys expect plugins to look like. It's written in TS (with tsdx), and I even made a simple unit test (for transformer, not for namer).
